### PR TITLE
Add GroupSharedLimit attribute support for Mesh, Amp and Node shaders

### PR DESCRIPTION
**Pull request by JoeCitizen**
- Adds support for the GroupSharedLimit feature for Mesh, Amplification and Node shaders.
- Tests for each of those shader types

### CODE REVIEW

File: tools/clang/test/HLSLFileCheckLit/hlsl/entry/attributes/GroupSharedLimitAmplification.hlsl [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+    Payload pld;
+    pld.dummy = float2(1.0, 2.0);
+    pld.pos = float4(g_testBufferFAIL[0], g_testBufferFAIL[1], g_testBufferFAIL[2], g_testBufferFAIL[3]);
+    pld.color[0] = 7.0;
+    pld.color[1] = 8.0;
+    DispatchMesh(NUM_THREADS, 1, 1, pld);
+}
+
+// RUN: not %dxc -E MainFail2 -T as_6_10 %s 2>&1 | FileCheck %s --check-prefix=CHECK-ERROR
```
Comment (damyanp) [inline tools/clang/test/HLSLFileCheckLit/hlsl/entry/attributes/GroupSharedLimitAmplification.hlsl]: I've been looking at this for a while, and I don't know if this is a common pattern in these sorts of tests.

As far as I can tell, MainFail and MainFail2 are identical apart from one having GroupSharedLimit and one not having it? But they're both expected to fail with the exact same error?

If this is the case, then:
* is there some coverage missing for the case when the specified GroupSharedLimit is different from the default?
* could we use something like the below to consolidate the identical code so the reader doesn't have to manually diff it understand the difference?

```
// RUN: not %dxc -E MainFail -T ......
// RUN: not %dxc -DUSE_GROUPSHARED_ATTRIBUTE -E MainFail -T ...
// CHECK-ERROR: Total Thread Group ....
#ifdef USE_GROUPSHARED_ATTRIBUTE
[GroupSharedLimit(NUM_BYTES_OF_SHARED_MEM)]
#endif
... etc ...
```

If my assumption about these being otherwise identical is incorrect then maybe a comment hinting at what the reader should be looking for would help?

Comment (JoeCitizen) [reply tools/clang/test/HLSLFileCheckLit/hlsl/entry/attributes/GroupSharedLimitAmplification.hlsl]: MainFail tests that we correctly check the limit when GroupSharedLimit is present, MainFail2 checks that the existing check (pre-SM 6.10) still works when you don't specify GroupSharedLimit

Comment (tex3d) [reply tools/clang/test/HLSLFileCheckLit/hlsl/entry/attributes/GroupSharedLimitAmplification.hlsl]: I agree with Damyan. This doesn't test setting the limit to something greater than or less than the default limit. This same problem exists in the node tests, and partly exists in the mesh test. Are they a problem in existing compute tests too? I'd have to check.

I think it would be easier to use arguments to specify the limit and size of the array accessed, like: `-DGROUPSHARED_LIMIT=16384 -DGROUPSHARED_BYTES=16388`, and a FileCheck argument like `--check-prefix=FAIL-16388`. Then you can have a set of run lines with the CHECK (for pass) and FAIL-### lines for a set of scenarios, with just one set of shader code following these.

Also, a more minimal shader would be sufficient. This is just testing the compiler, so you just need to make sure the groupshared memory is used with an unknown dynamic index that's preserved in final DXIL.

This makes it easier to test a wider variety of scenarios with minimal effort.

Example:
```hlsl
// RUN: %dxc -T as_6_10 %s 2>&1 | FileCheck %s --check-prefix=PASS
// RUN: %dxc -T as_6_10 -DGROUPSHARED_BYTES=32768 %s 2>&1 | FileCheck %s --check-prefix=PASS
// RUN: not %dxc -T as_6_10 -DGROUPSHARED_BYTES=32772 %s 2>&1 | FileCheck %s --check-prefix=ERROR-32772

// RUN: %dxc -T as_6_10 -DGROUPSHARED_LIMIT=0 %s 2>&1 | FileCheck %s --check-prefix=PASS
// RUN: %dxc -T as_6_10 -DGROUPSHARED_LIMIT=16384 %s 2>&1 | FileCheck %s --check-prefix=PASS
// RUN: %dxc -T as_6_10 -DGROUPSHARED_BYTES=16384 -DGROUPSHARED_LIMIT=16384 %s 2>&1 | FileCheck %s --check-prefix=PASS
// RUN: %dxc -T as_6_10 -DGROUPSHARED_BYTES=16380 -DGROUPSHARED_LIMIT=16384 %s 2>&1 | FileCheck %s --check-prefix=PASS
// RUN: not %dxc -T as_6_10 -DGROUPSHARED_BYTES=16388 -DGROUPSHARED_LIMIT=16384 %s 2>&1 | FileCheck %s --check-prefix=ERROR-16388

// PASS: @main
// ERROR-32772: Total Thread Group Shared Memory storage is 32772, exceeded 32768.
// ERROR-16388: Total Thread Group Shared Memory storage is 16388, exceeded 16384.

struct Payload { uint u; };

#ifdef GROUPSHARED_BYTES
#define GROUPSHARED_DWORDS (GROUPSHARED_BYTES / 4)
groupshared uint g_TestBuffer[GROUPSHARED_DWORDS];
#endif

#ifdef GROUPSHARED_LIMIT
[GroupSharedLimit(GROUPSHARED_LIMIT)]
#endif
[numthreads(32, 1, 1)]
void main(in uint tid : SV_DispatchThreadID)
{
    Payload pld = (Payload)0;
#ifdef GROUPSHARED_BYTES
    g_TestBuffer[tid] = tid;
    GroupMemoryBarrierWithGroupSync();
    pld.u = g_TestBuffer[WaveActiveMax(g_TestBuffer[tid])];
#endif
    DispatchMesh(1, 1, 1, pld);
}
```

File: tools/clang/test/HLSLFileCheckLit/hlsl/entry/attributes/GroupSharedLimitAmplification.hlsl [hunk 2]
```diff
@@ -0,0 +1,54 @@
+// REQUIRES: dxil-1-10
+
+// Pass: GSM fits within the GroupSharedLimit
+// RUN: %dxc -E main -T as_6_10 -DGSM_EXTRA=0 -DUSE_GROUP_SHARED_LIMIT %s | FileCheck %s
+// CHECK: @main
+
+// Fail: GSM exceeds GroupSharedLimit (32772 > 32768)
```
Comment (tex3d) [inline tools/clang/test/HLSLFileCheckLit/hlsl/entry/attributes/GroupSharedLimitAmplification.hlsl]: In this case, the limit is always the same as the default.

You're testing:
- FAIL: default == limit < usage
- FAIL: default < usage (no limit)

You're not testing:
- PASS: usage <= default (no limit)
- PASS: default < usage <= limit
- PASS: no usage, limit=0 (edge case)
- PASS: limit == usage < default
- FAIL: limit < usage < default
- FAIL: limit=0 < usage < default (edge case)

My suggestion still seems better than this approach just using `GSM_EXTRA` and `USE_GROUP_SHARED_LIMIT` flags, plus it could easily be adapted to other shader model entries.

File: tools/clang/test/HLSLFileCheckLit/hlsl/entry/attributes/GroupSharedLimitMesh.hlsl [hunk 3]
```diff
@@ -0,0 +1,76 @@
+// REQUIRES: dxil-1-10
+
+// Pass: GSM fits within the GroupSharedLimit
```
Comment (tex3d) [inline tools/clang/test/HLSLFileCheckLit/hlsl/entry/attributes/GroupSharedLimitMesh.hlsl]: These test:
- PASS: usage == limit == default
- FAIL: default < limit < usage
- FAIL: default < usage (no limit)

But not:
- PASS: usage <= default (no limit)
- PASS: default < usage <= limit
- PASS: no usage, limit=0 (edge case)
- PASS: limit == usage < default
- FAIL: limit < usage < default
- FAIL: limit=0 < usage < default (edge case)

File: tools/clang/test/HLSLFileCheckLit/hlsl/entry/attributes/GroupSharedLimitNode.hlsl [hunk 4]
```diff
@@ -0,0 +1,46 @@
+// REQUIRES: dxil-1-10
+// RUN: %dxc -T lib_6_10 %s | FileCheck %s
+
+// Node shader test for GroupSharedLimit attribute
+// This tests that the GroupSharedLimit attribute is correctly applied to node shaders
+// and that the metadata is emitted correctly.
```
Comment (tex3d) [inline tools/clang/test/HLSLFileCheckLit/hlsl/entry/attributes/GroupSharedLimitNode.hlsl]: This tests:
- PASS: usage == limit == default

But not:
- PASS: usage <= default (no limit)
- PASS: default < usage <= limit
- PASS: no usage, limit=0 (edge case)
- PASS: limit == usage < default

Plus, failure cases could be integrated into one test if you used a suggested pattern.

File: tools/clang/test/HLSLFileCheckLit/hlsl/entry/attributes/GroupSharedLimitNodeError.hlsl [hunk 5]
```diff
@@ -0,0 +1,46 @@
+// REQUIRES: dxil-1-10
+// RUN: not %dxc -T lib_6_10 %s 2>&1 | FileCheck %s
+
+// Node shader error test for GroupSharedLimit attribute
+// This test verifies that exceeding the GroupSharedLimit produces an error.
+
+// CHECK: Total Thread Group Shared Memory storage is 32772, exceeded 32768.
```
Comment (tex3d) [inline tools/clang/test/HLSLFileCheckLit/hlsl/entry/attributes/GroupSharedLimitNodeError.hlsl]: This tests:
- FAIL: default == limit < usage

But not:
- FAIL: default < limit < usage
- FAIL: default < usage (no limit)
- FAIL: limit < usage < default
- FAIL: limit=0 < usage < default (edge case)

File: tools/clang/test/HLSLFileCheckLit/hlsl/entry/attributes/GroupSharedLimit.hlsl [hunk 6]
```diff
@@ -1,77 +1,41 @@
 // REQUIRES: dxil-1-10
-// RUN: %dxc -E MainPass -T cs_6_10 %s | FileCheck %s
+
+// Pass: GSM fits within the GroupSharedLimit
```
Comment (tex3d) [inline tools/clang/test/HLSLFileCheckLit/hlsl/entry/attributes/GroupSharedLimit.hlsl]: Updated test checks:
- PASS: usage == limit == default
- FAIL: default == limit < usage
- FAIL: default < usage (no limit)

You're not testing:
- PASS: usage <= default (no limit)
- PASS: default < usage <= limit
- PASS: no usage, limit=0 (edge case)
- PASS: limit == usage < default
- FAIL: limit < usage < default
- FAIL: limit=0 < usage < default (edge case)

File: lib/DxilValidation/DxilValidation.cpp [hunk 7]
```diff
@@ -3959,17 +3959,47 @@ static void ValidateGlobalVariables(ValidationContext &ValCtx) {
   }
 
   // Check if the entry function has attribute to override TGSM size.
-  if (M.HasDxilEntryProps(M.GetEntryFunction())) {
+  // For library profiles, iterate over all entry functions and validate
+  // TGSM against each entry's groupSharedLimitBytes since TGSM is global.
+  if (M.GetShaderModel()->IsLib()) {
```
Comment (tex3d) [inline lib/DxilValidation/DxilValidation.cpp]: Digging into this, I think the diagnostic part needs improvement, but there are multiple issues.

Basically, the issues are:

1. TGSMSize is the total size of all TGSM globals declared in a library, which isn't correct for library entry validation. We need a different approach that counts per-entry TGSM usage, and accounts for called functions.

2. The original diagnostic is being used beyond the original intent. If based on a custom limit, I think there should be a different diagnostic, otherwise it's unclear where that limit came from.

For some reason two rules: kMaxTGSMSize and SmMaxMSSMSize have identical text! No reason to split these in the first place. Instead, we can split this by default or user-specified limit. Like: "exceeds maximum of %1 bytes" or "exceeds specified limit of %1 bytes".

3. Iterating backward and reporting just the last GV made some sense when it was all one shader with one list of GVs that were all used by that shader, but it doesn't make sense from a particular library entry perspective which might not even reference that specific GV.

This diagnostic could report the first use inside the entry point of the GV that pushes that entry over the size limit. This would also make the library case a different diagnostic/wording than the global entry case.

The following code collects TGSM usage properly for functions in a library and can be adapted to use for diagnostics:
https://github.com/microsoft/DirectXShaderCompiler/blob/3e6e148537683c22e3e74977d56516f16f39c7be/lib/DxilContainer/DxilContainerAssembler.cpp#L1455-L1490

Here's my adaptation for DxilValidation in a patch:
https://gist.github.com/tex3d/bff961303eb8834033e5ec95f2b0d422

Assumes new diagnostic rules, and not built/tested, but should get you 90% of the way there.

File: tools/clang/lib/CodeGen/CGHLSLMS.cpp [hunk 8]
```diff
@@ -1655,8 +1655,9 @@ void CGMSHLSLRuntime::AddHLSLFunctionInfo(Function *F, const FunctionDecl *FD) {
   } else {
     if (SM->IsMS()) { // Fallback to default limits
       funcProps->groupSharedLimitBytes = DXIL::kMaxMSSMSize; // 28k For MS
-    } else if (SM->IsAS() || SM->IsCS()) {
-      funcProps->groupSharedLimitBytes = DXIL::kMaxTGSMSize; // 32k For AS/CS
+    } else if (SM->IsAS() || SM->IsCS() || funcProps->IsNode()) {
```
Comment (tex3d) [inline tools/clang/lib/CodeGen/CGHLSLMS.cpp]: We treat `-1` as no user-specified limit, right? Can we just set it to `-1` here?

File: lib/DxilValidation/DxilValidation.cpp (hunk 7 above)
Comment (tex3d) [reply lib/DxilValidation/DxilValidation.cpp]: Ok, I updated things to build, fixed up the diagnostics, and fixed up tests. Uploaded the branch to my fork here: 

It merges some recent changes from main as well, so just [diffing the branches](https://github.com/JoeCitizen/DirectXShaderCompiler/compare/Variable-Shared-Memory-Fixes...tex3d:DirectXShaderCompiler:Variable-Shared-Memory-Fixes) isn't clean.

You can cleanly fast-forward your branch to mine. If you want me to push the changes to this branch, just give the word.

File: lib/DxilValidation/DxilValidation.cpp [hunk 9]
```diff
@@ -3916,70 +3916,156 @@ static void ValidateGlobalVariables(ValidationContext &ValCtx) {
...
+  auto getMaxTGSM = [](const DxilFunctionProps &Props) {
+    unsigned MaxSize = 0;
+    if (Props.IsCS() || Props.IsAS() || Props.IsNode())
+      MaxSize = DXIL::kMaxTGSMSize;
+    else if (Props.IsMS())
+      MaxSize = DXIL::kMaxMSSMSize;
+    if (Props.groupSharedLimitBytes >= 0)
+      MaxSize = static_cast<unsigned>(Props.groupSharedLimitBytes);
+    return MaxSize;
+  };
```
Comment (damyanp) [inline lib/DxilValidation/DxilValidation.cpp]: It took me a little while to figure out the intent of this. I think groupSharedLimitByes always overrides everything?

Is the case when it would return 0 an expected one or is that actually a sign of a bug?

```suggestion
  auto getMaxTGSM = [](const DxilFunctionProps &Props) {
    if (Props.groupSharedLimitBytes >= 0)
      return static_cast<unsigned>(Props.groupSharedLimitBytes);
    if (Props.IsCS() || Props.IsAS() || Props.IsNode())
      return DXIL::kMaxTGSMSize;
    else if (Props.IsMS())
      return DXIL::kMaxMSSMSize;
    return 0;
  };
```

File: include/dxc/DXIL/DxilFunctionProps.h [hunk 10]
```diff
@@ -117,7 +117,7 @@ struct DxilFunctionProps {
     memset(&Node, 0, sizeof(Node));
     Node.LaunchType = DXIL::NodeLaunchType::Invalid;
     Node.LocalRootArgumentsTableIndex = -1;
-    groupSharedLimitBytes = 0;
+    groupSharedLimitBytes = -1;
```
Comment (V-FEXrt) [inline include/dxc/DXIL/DxilFunctionProps.h]: I won't block on this, but I feel that using magic numbers to encode information is generally a bad idea.

If you want to encode so notion of "not-set" using something like std::optional is more appropriate

File: lib/DXIL/DxilMetadataHelper.cpp [hunk 11]
```diff
@@ -1626,7 +1626,7 @@ MDTuple *DxilMDHelper::EmitDxilEntryProperties(uint64_t rawShaderFlag,
     }
 
     const hlsl::ShaderModel *SM = GetShaderModel();
-    if (SM->IsSMAtLeast(6, 10)) {
+    if (SM->IsSMAtLeast(6, 10) && props.groupSharedLimitBytes != -1) {
```
Comment (V-FEXrt) [inline lib/DXIL/DxilMetadataHelper.cpp]: More important than using std::optional, this should be named in some manner[1]. What does it _mean_ that grouSharedLimitBytes is `-1`?


[1]: my personal preference would be a tiny helper function but I'm flexible. Another option is defining as constant for -1 with a name that gives enough context to the meaning

File: utils/hct/hctdb.py [hunk 12]
```diff
@@ -8879,9 +8879,15 @@ def build_valrules(self):
...
-            "Total Thread Group Shared Memory storage is %0, exceeded %1.",
+        self.add_valrule_msg(
+            "Sm.MaxTGSMSizeOnEntry",
+            "Total Thread Group Shared Memory used by entry must not exceed maximum for shader model",
+            "Total Thread Group Shared Memory used by '%0' is %1, exceeding maximum: %2.",
+        )
+        self.add_valrule_msg(
+            "Sm.ExplicitTGSMSizeOnEntry",
+            "Total Thread Group Shared Memory used by entry must not exceed limit specified by entry attribute",
+            "Total Thread Group Shared Memory used by '%0' is %1, exceeding explicit limit: %2.",
```
Comment (damyanp) [inline utils/hct/hctdb.py:8890]: Should the messages end with a period?


```suggestion
        self.add_valrule_msg(
            "Sm.MaxTGSMSizeOnEntry",
            "Total Thread Group Shared Memory used by entry must not exceed maximum for shader model.",
            "Total Thread Group Shared Memory used by '%0' is %1, exceeding maximum: %2.",
        )
        self.add_valrule_msg(
            "Sm.ExplicitTGSMSizeOnEntry",
            "Total Thread Group Shared Memory used by entry must not exceed limit specified by entry attribute.",
            "Total Thread Group Shared Memory used by '%0' is %1, exceeding explicit limit: %2.",
```

File: include/dxc/DXIL/DxilFunctionProps.h (hunk 10 above)
Comment (V-FEXrt) [reply include/dxc/DXIL/DxilFunctionProps.h]: I think I've now deduced that -1 -> Use max size? That seems a bit counterintuitive to me but probably a named constant is enough to clear that up

**Review by V-FEXrt — COMMENTED** [review]
Generally LGTM but the more I sat with it, the more I think we need to do something about the `-1` magic. Happy to leave it up to you on exactly what to do but it should be named in some manner

File: include/dxc/DXIL/DxilFunctionProps.h (hunk 10 above)
Comment (JoeCitizen) [reply include/dxc/DXIL/DxilFunctionProps.h]: I generally I agree, I was following the convention that appears to be set in the line above where LocalRootArgumentsTableIndex is initialised to -1

File: utils/hct/hctdb.py (hunk 12 above)
Comment (tex3d) [reply utils/hct/hctdb.py:8890]: I think the majority of rule descriptions (using `add_valrule_msg`) do not, but there's no consistency, so we can add periods to these.

File: include/dxc/DXIL/DxilFunctionProps.h (hunk 10 above)
Comment (tex3d) [reply include/dxc/DXIL/DxilFunctionProps.h]: Generally, we either need a special value for "unset" or an additional flag somewhere. We can add a define for this to make the code semantically better, or we can add another field/flag. Adding a flag won't change or require more metadata because this should be preserved by whether the extended metadata tag is present at all, not by any special value. We could even add accessors (get/set/isSet/unset) to keep the flag in sync. Though just adding a special defined value for "unset" seems like the simplest solution.

Comment (tex3d) [reply include/dxc/DXIL/DxilFunctionProps.h]: @V-FEXrt: -1 doesn't mean "max size", it means the default isn't overridden for the shader entry. The max size used when not overridden by the shader is a long-standing default value from defines (32k for compute-like shaders except mesh shader which is 28k). The shader can set a larger size than that default, and if it uses more than the default, the runtime will check device capability to make sure the shader can run.

Comment (tex3d) [reply include/dxc/DXIL/DxilFunctionProps.h]: Oh yeah, given this definition, I think `std::optional` would work fine instead of an explicit flag and extra accessors defined in this class.

Comment (V-FEXrt) [reply include/dxc/DXIL/DxilFunctionProps.h]: Makes sense, to be clear, I'm not specifically set on `std:optional`, I would be happy with a simple `#define`, we just have a lot of `-1`s floating around in the PR. even something like `GS_USE_DEFAULT_SIZE` would make the later code a lot easier to read

File: lib/DxilValidation/DxilValidation.cpp (hunk 9 above)
Comment (tex3d) [reply lib/DxilValidation/DxilValidation.cpp]: Returning `0` is an intended case, yes. Your code looks cleaner and will do the same thing, so I'll just accept it.
